frame/assets/src/lib.rs

```diff
@@ -247,19 +256,584 @@ decl_error! {
+		) {
+			T::ForceOrigin::ensure_origin(origin)?;
+			let owner = T::Lookup::lookup(owner)?;
+
+			ensure!(!Asset::<T>::contains_key(id), Error::<T>::InUse);
+
+			Asset::<T>::insert(id, AssetDetails {
+				owner: owner.clone(),
+				issuer: owner.clone(),
+				admin: owner.clone(),
+				freezer: owner.clone(),
+				supply: Zero::zero(),
+				deposit: Zero::zero(),
+				max_zombies,
+				min_balance,
+				zombies: Zero::zero(),
+				accounts: Zero::zero(),
+			});
+			Self::deposit_event(RawEvent::ForceCreated(id, owner));
+		}
+
+		/// Destroy a class of fungible assets owned by the sender.
+		///
+		/// The origin must be Signed and the sender must be the owner of the asset `id`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		/// asset.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::destroy()]
+		fn destroy(origin, #[compact] id: T::AssetId) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Asset::<T>::try_mutate_exists(id, |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+				ensure!(details.owner == origin, Error::<T>::NoPermission);
+				ensure!(details.accounts == details.zombies, Error::<T>::RefsLeft);
+				T::Currency::unreserve(&details.owner, details.deposit);
+
+				*maybe_details = None;
+				Account::<T>::remove_prefix(&id);
+				Self::deposit_event(RawEvent::Destroyed(id));
+				Ok(())
+			})
+		}
+
+		/// Destroy a class of fungible assets.
+		///
+		/// The origin must conform to `ForceOrigin`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		/// asset.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::force_destroy()]
+		fn force_destroy(origin, #[compact] id: T::AssetId) -> DispatchResult {
+			T::ForceOrigin::ensure_origin(origin)?;
+
+			Asset::<T>::try_mutate_exists(id, |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::Unknown)?;
+				ensure!(details.accounts == details.zombies, Error::<T>::RefsLeft);
+				T::Currency::unreserve(&details.owner, details.deposit);
+
+				*maybe_details = None;
+				Account::<T>::remove_prefix(&id);
+				Self::deposit_event(RawEvent::Destroyed(id));
+				Ok(())
+			})
+		}
+
+		/// Mint assets of a particular class.
 		///
-		/// TWOX-NOTE: `AssetId` is trusted, so this is safe.
-		TotalSupply: map hasher(twox_64_concat) T::AssetId => T::Balance;
+		/// The origin must be Signed and the sender must be the Issuer of the asset `id`.
+		///
+		/// - `id`: The identifier of the asset to have some amount minted.
+		/// - `beneficiary`: The account to be credited with the minted assets.
+		/// - `amount`: The amount of the asset to be minted.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(1)`
+		/// Modes: Pre-existence of `beneficiary`
```

Comment:
what are `Modes` ? the worst case of function for computing the weight ?
if so I think reading beneficiary existence and writing beneficiary ref count increment is the worst case

Reply:
Not the worst case, but the conditions which dictate the cases.

Reply:
ah ok 👍

Reply:
(Updated)